File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/core/data/VelaStore.kt
package app.getvela.wallet.core.data

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first

/**
 * The wallet's key-value space on this device: string keys, string values, one
 * file.
 *
 * **Why one.** `vela.serviceEndpoints` already has two readers — the onboarding
 * endpoint sheet and, from spec 040, the `network_admin` machine's settings
 * pages. Two DataStore files holding the same key name is how a person changes
 * their bundler in Settings and onboarding keeps using the old one, with
 * nothing in either code path looking wrong.
 *
 * **Why strings.** These keys are the *cross-client compatibility contract*:
 * the same names and the same camelCase JSON payloads the web and desktop
 * clients read and write, so a wallet's address book follows its owner between
 * devices. Typed accessors would invite reshaping on the way through, and a
 * reshaped record is a record the other clients no longer recognise. Values go
 * in and come out as the bytes the core produced.
 *
 * **What belongs here**: anything keyed `vela.*` that survives a restart and
 * belongs to the *account* — contacts, groups, networks, endpoints, display
 * preferences. **What does not**: the theme preference, which lives in its own
 * store precisely so that signing out cannot reach it.
 *
 * Reads and writes are best-effort by design (spec 040
 * `contracts/shell-operations.md`): a machine asked to save must still be
 * *answered* when saving fails, or it waits forever on an effect nobody will
 * resolve. So nothing here throws for a storage fault; the caller learns the
 * outcome from the return value.
 *
 * **Why an interface with one shipping implementation.** The second
 * implementation is the test one. Every executor in this app is a function from
 * an operation to an answer *over storage*, and the whole point of the unit
 * suite is to check those answers without a device — including the answers for
 * storage that fails, which a real DataStore will not produce on request.
 */
interface KeyValueStore {

    /** The stored value, or `null` when the key was never written. */
    suspend fun read(key: String): String?

    /** Write one key. Returns `false` when the platform refused. */
    suspend fun write(key: String, value: String): Boolean

    /**
     * Remove keys, in one edit.
     *
     * Together, because a caller removing two related keys means both or
     * neither: signing out leaves a device with an account list gone and an
     * active index pointing into it if only the first removal lands.
     */
    suspend fun remove(vararg keys: String): Boolean

    /** The keys spec 040's machines own. Names are the contract — see [VelaStore]. */
    object Keys {
        const val CONTACTS = "vela.contacts"
        const val CONTACTS_DISMISSED = "vela.contacts.dismissed"
        const val CONTACT_GROUPS = "vela.contactGroups"
        const val DISPLAY_CURRENCY = "vela.displayCurrency"
        const val CUSTOM_NETWORKS = "vela.customNetworks"

        /**
         * Singular — `networkConfig`, not `networkConfigs`.
         *
         * Verified against `app-web/.../settings/core/network-admin-executor.ts`,
         * where the same key is written. A plural here would be a store the
         * other clients cannot read, and nothing would report it: the settings
         * page would simply show no overrides.
         */
        const val NETWORK_CONFIG = "vela.networkConfig"
        const val RPC_PROVIDERS = "vela.rpcProviders"

        /** Shared with onboarding's endpoint sheet — the reason this class exists. */
        const val SERVICE_ENDPOINTS = "vela.serviceEndpoints"

        /**
         * Endpoints the pool has ruled against, as `{url, bannedAt, permanent}`
         * (spec 041). Shared with the other clients: a device that learned an
         * endpoint is dead should agree with its siblings rather than each
         * rediscovering it the hard way.
         */
        const val RPC_BANNED = "vela.rpc.banned"

        /** `address → {usd, at}`, 24h TTL — the Expo bytes (spec 041). */
        const val BALANCE_CACHE = "vela.balanceCache"

        /** Whether money figures are hidden on this device. */
        const val BALANCE_HIDDEN = "vela.balanceHidden"
    }
}

/** The shipping [KeyValueStore]: one DataStore file, shared by every machine. */
class VelaStore(private val context: Context) : KeyValueStore {

    override suspend fun read(key: String): String? =
        runCatching { context.velaStore.data.first()[stringPreferencesKey(key)] }.getOrNull()

    override suspend fun write(key: String, value: String): Boolean =
        runCatching { context.velaStore.edit { it[stringPreferencesKey(key)] = value } }.isSuccess

    override suspend fun remove(vararg keys: String): Boolean = runCatching {
        context.velaStore.edit { preferences ->
            keys.forEach { preferences.remove(stringPreferencesKey(it)) }
        }
    }.isSuccess
}

/**
 * The wallet's DataStore file.
 *
 * The name is `vela_onboarding` for compatibility, not for accuracy: it is what
 * every installed build already keeps its accounts in, and renaming it would
 * sign every existing person out. `AccountStore` opens the same file by the
 * same name.
 */
internal val Context.velaStore: DataStore<Preferences> by preferencesDataStore(
    name = "vela_onboarding",
)

File: app-ios/VelaWallet/VelaWallet/Core/VelaStore.swift
//
//  VelaStore.swift
//  VelaWallet
//
//  The device's `vela.*` key-value store — the cross-client compatibility seam.
//
//  This is `AccountStore`'s storage half, generalised so the wallet-state
//  machines can reach the same shelf the account list already lives on. It adds
//  no policy: the account rules stay in `AccountStore`, the business rules stay
//  in `vela-core`, and this file only knows how to put JSON text under a name.
//
//  ## Values are JSON TEXT, and that is the contract
//
//  Not a plist array, not an archived object — the exact string web writes to
//  `localStorage` and the Expo client writes to `AsyncStorage`. That is what
//  makes "a contact written on the web reads on the phone" literally true
//  rather than approximately true, and it is why every accessor here goes
//  through `JSONSerialization` instead of `UserDefaults`' typed getters.
//
//  ## Corrupt reads answer empty — they never throw
//
//  A damaged blob must not make a screen permanently unopenable, and it must
//  not be reported to the core as a failure either: `vela-core`'s own operation
//  docs assume it ("Unreadable/corrupt answers as empty, exactly as the TS
//  loaders' `catch { [] }` does"). A machine that receives a *failed* store read
//  stays unloaded forever and silently drops every write that follows.
//
//  ## Records come out whole
//
//  The accessors deal in `[String: Any]`, deliberately. `AccountStore`'s module
//  doc explains what a typed struct costs here: a mapper that copies an account
//  field by field drops `keys`, and the next restore derives a DIFFERENT Safe
//  address. Views get decoded into types (`CoreViews.swift`); records that go
//  back out to disk or to the core are carried, not rebuilt.
//

import Foundation

struct VelaStore {

    /// Every `vela.*` name this app reads or writes, in one place.
    ///
    /// Centralised because they are a contract with three other clients rather
    /// than this app's private business — a typo here is not a local bug, it is
    /// a wallet that cannot see the address book it wrote yesterday. The
    /// capitalised initialisms are load-bearing too: shipped data says
    /// `rpcURL`, not `rpcUrl`.
    enum Key {
        // Owned by onboarding (spec 019).
        static let accounts = "vela.accounts"
        static let activeIndex = "vela.activeAccountIndex"
        static let pendingUploads = "vela.pendingUploads"

        // Owned by `contacts` (spec 050).
        static let contacts = "vela.contacts"
        static let contactsDismissed = "vela.contacts.dismissed"
        static let contactGroups = "vela.contactGroups"

        // Owned by `network_admin` (spec 050).
        static let customNetworks = "vela.customNetworks"
        static let networkConfig = "vela.networkConfig"
        static let rpcProviders = "vela.rpcProviders"

        /// Written by BOTH `network_admin` and `AccountStore` — reach it only
        /// through `AccountStore`, never directly (data-model §5). Two
        /// independent writers is how a person's custom passkey-index endpoint
        /// disappears the first time they open 设置 → 端点.
        static let serviceEndpoints = "vela.serviceEndpoints"

        // Owned by `display_currency` (spec 050).
        static let displayCurrency = "vela.displayCurrency"

        // Owned by the read path (spec 051).
        /// `address → { usd, at }`, 24-hour TTL. The last total the wallet
        /// knew, so a cold start shows a figure instead of a spinner.
        static let balanceCache = "vela.balanceCache"
        /// The ERC-20s the wallet reads balances for.
        static let customTokens = "vela.customTokens"
        /// The local transaction store — the activity feed's single source of
        /// truth, capped at the newest 200 records on every client.
        static let transactionHistory = "vela.transactionHistory"
        /// USD → fiat, as Chainlink's mainnet feeds last stated it (5-minute
        /// TTL in memory; persisted so a cold or offline start still converts).
        static let fiatRates = "vela.fiatRates.v1"
        /// Those feeds' ENS-resolved addresses. The proxies are immutable, so
        /// this is a 30-day cache rather than a lookup on every launch.
        static let fiatFeedAddrs = "vela.fiatFeedAddrs.v1"
        /// The configurable endpoint's USD-based rate map, keyed by the URL it
        /// came from so changing the endpoint refetches.
        static let fxRates = "vela.fxRates.v1"
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    // MARK: - Lists

    /// Absent, unparseable, or not-an-array all read as `[]`.
    func readList(_ key: String) -> [[String: Any]] {
        json(key) as? [[String: Any]] ?? []
    }

    func writeList(_ key: String, _ value: [[String: Any]]) {
        write(key, value)
    }

    // MARK: - Objects

    /// Absent, unparseable, or not-an-object all read as `[:]`.
    func readObject(_ key: String) -> [String: Any] {
        json(key) as? [String: Any] ?? [:]
    }

    func writeObject(_ key: String, _ value: [String: Any]) {
        write(key, value)
    }

    // MARK: - Bare strings

    /// `nil` means the key was never written — which several machines read as a
    /// real answer ("the person never chose a currency"), distinct from any
    /// value the key could hold.
    /// Whether this key has ever been written.
    ///
    /// The difference between "nothing here" and "could not be read" is a real
    /// one for the send history: an empty list tells the core nobody has ever
    /// been paid, and it would then treat every address as a first
    /// interaction — the address-poisoning warning, shown to everyone, forever.
    func hasKey(_ key: String) -> Bool {
        defaults.object(forKey: key) != nil
    }

    func readString(_ key: String) -> String? {
        guard let raw = defaults.string(forKey: key), !raw.isEmpty else { return nil }
        return raw
    }

    func writeString(_ key: String, _ value: String?) {
        guard let value, !value.isEmpty else {
            defaults.removeObject(forKey: key)
            return
        }
        defaults.set(value, forKey: key)
    }

    func remove(_ key: String) {
        defaults.removeObject(forKey: key)
    }

    // MARK: - The one parse, the one write

    private func json(_ key: String) -> Any? {
        guard let raw = defaults.string(forKey: key),
              let data = raw.data(using: .utf8)
        else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }

    /// A value that cannot be serialized is DROPPED rather than written badly.
    ///
    /// Writing half a record, or writing `"[object Object]"`, would corrupt a
    /// shelf three other clients read. Keeping yesterday's good bytes is the
    /// better failure: the core's ledger is authoritative in memory, so the
    /// person's session is unaffected and the next successful write repairs it.
    private func write(_ key: String, _ value: Any) {
        guard JSONSerialization.isValidJSONObject(value),
              let data = try? JSONSerialization.data(withJSONObject: value),
              let text = String(data: data, encoding: .utf8)
        else { return }
        defaults.set(text, forKey: key)
    }
}
